File: src/app/hooks/overview.ts
import { Octokit } from '@octokit/rest';
import * as dotenv from 'dotenv';

// 加载环境变量
dotenv.config();

// 定义类型
interface RepoInfo {
    stars: number;
    forks: number;
    commits: number;
}

// 初始化 GitHub 客户端
const octokit = new Octokit({
    auth: process.env.GITHUB_TOKEN,
});

// 获取仓库信息
export async function getOverview(repoName: string): Promise<RepoInfo | null> {
    if (!repoName) {
        console.error('Repository name is required');
        return null;
    }

    try {
        // 分割仓库名（格式：owner/repo）
        const [owner, repo] = repoName.split('/');
        if (!owner || !repo) {
            console.error('Invalid repository name format. Expected "owner/repo"');
            return null;
        }

        // 获取仓库基本信息
        const { data: repoData } = await octokit.rest.repos.get({ owner, repo });

        // 获取提交总数
        const commitCount = await getCommitCount(owner, repo);

        const info: RepoInfo = {
            stars: repoData.stargazers_count,
            forks: repoData.forks_count,
            commits: commitCount,
        };

        return info;
    } catch (error) {
        console.error(`An error occurred: ${error}`);
        return null;
    }
}


async function getCommitCount(owner: string, repo: string): Promise<number> {
    let totalCommits = 0;
    let page = 1;

    try {
        while (true) {
            // 等待 100ms，防止被 GitHub 限流
            // await new Promise(resolve => setTimeout(resolve, 10));

            const { data: commits, headers } = await octokit.rest.repos.listCommits({
                owner,
                repo,
                per_page: 500,
                page,
            });

            if (commits.length === 0) break;

            totalCommits += commits.length;

            const linkHeader = headers.link;
            if (!linkHeader || !linkHeader.includes('rel="next"')) break;

            page++;
        }
    } catch (error) {
        console.error(`Failed to fetch commit count for ${owner}/${repo}:`, error);
        return 0;
    }

    return totalCommits;
}

File: server/service/overview.py
import os
from typing import Optional
from github import Github
# 使用环境变量存储 token
TOKEN = os.getenv("GITHUB_TOKEN")  # 设置环境变量: export GITHUB_TOKEN=ghp_xxx...
g = Github(TOKEN) if TOKEN else Github()  # 有 token 就用，没有就匿名（不推荐生产使用）


def get_overview(repo_name: Optional[str] = None):
    if not repo_name:
        return None

    try:
        repo = g.get_repo(repo_name)

        # 注意：.totalCount 在 PyGithub 中可能不会立即生效（尤其是大仓库），可用 len(list()) 替代（但更慢）
        # 所以这里我们只取最新一页来估算或精确获取（如果 commit 不多）
        commits = repo.get_commits()
        # 如果只想快速获取数量而不在乎完全准确，可以用 totalCount（基于 API 返回头）
        commit_count = commits.totalCount  # 这个值是准确的，不需要遍历所有页

        info = {
            "stars": repo.stargazers_count,
            "forks": repo.forks_count,
            "commits": commit_count,
        }

        return info

    except Exception as e:
        print(f"An error occurred while fetching {repo_name}: {e}")
        return None

    finally:
        # 好习惯：及时关闭连接
        g.close()
